README.md
[![Docker Repository on Quay](https://quay.io/repository/lifechurch/k8s-deploy-helper/status "Docker Repository on Quay")](https://quay.io/repository/lifechurch/k8s-deploy-helper)

# Description
k8s-deploy-helper (KDH) is a tool to help build and deploy containerized applications into Kubernetes using GitLab CI along with templated manifest files. Major features include:

* Automated Kubernetes Secret Management using GitLab's UI. Version 3.0.0 can automatically insert secrets into Kubernetes manifests and Dockerfiles to lessen manual work.
* Build via Heroku buildpacks or via Dockerfiles starting in 3.1.0.
* Automated canary deployments. Dynamic creation is included in version 3.0.0.
* Automated review app deployments.
* Automated deployment of applications using Heroku Procfile in 3.1.0.
* Deployment notifications to New Relic, Datadog and Slack.
* Templated manifest deployments to Kubernetes living in the same repo as the code, giving developers more control.
* Uses [kubeval](https://github.com/garethr/kubeval) to evaluate manifest yaml before any are deployed.
* Easy, standardized image creation with build arguments and multiple Dockerfile support.
* Standardized building conventions to allow for easy rollbacks through GitLab UI and better caching.


This project is not endorsed or affiliated with GitLab in any way.

# Examples
In addition to this documentation, the best way to get started is to look at our [example repository](https://github.com/lifechurch/example-go).

Need some help getting started? Feel free to join us on [Open Digerati Slack](https://join.slack.com/t/opendigerati/shared_invite/enQtMjU4MTcwOTIxMzMwLTcyYjQ4NWEwMzBlOGIzNDgyM2U5NzExYTY3NmI0MDE4MTRmMTQ5NjNhZWEyNDY3N2IyOWZjMDIxM2MwYjEwMmQ) in #k8s and we'll be more than happy to assist.

# Why?
GitLab's Auto DevOps initiative is amazing for getting simple apps running quickly, but for slightly more complex and production workloads, you need more control in the process. For instance, what if you have a pod with sidecar containers? What if you want a deployment of worker pods using something like celery for async work? You'll need to interact with Kubernetes at a deeper level to do stuff like this, and that's where our project comes in.

At Life.Church, we wanted to create a standardized tool along with corresponding conventions that our developers could use with GitLab CI to allow us to get up and running with manifest-based deployments as quickly and easily as possible. So, we took the work that GitLab started, and used it as the base of a new project that would meet our needs.

This tool was built akin to an airplane that was built while we were flying it. Our process is constantly maturing as we learn more about how to deploy into k8s. Our goal isn't to say 'this is the best way of deploying', but simply to share how we're doing it now, knowing that it will at least be a helpful starting place for others who are embarking on their Kubernetes journey.

# Prerequisites

* GitLab w/customizable runners
* Kubernetes

## Configuring GitLab Runner

There is a lot of discussion around the best way to build docker images from within Docker. We ended up going the route of sharing the Docker socket.  Here is a sample GitLab Runner configuration. Of particular note, is the volumes section, to share the socket the way we expect. Additionally, we use dind for some stages, so privileged needs to be turned on as well.

```
[[runners]]
  name = "runner01"
  limit = 16
  url = "https://runneriurl/ci"
  token = "token"
  executor = "docker"
  [runners.docker]
    tls_verify = false
    image = "docker:latest"
    privileged = true
    disable_cache = false
    volumes = ["/var/run/docker.sock:/var/run/docker.sock", "/cache"]
    shm_size = 0
  [runners.cache]
```

## Integrate Kubernetes into your Project

In your GitLab Project, go to Operations->Kubernetes to give GitLab the ability to talk to your Kubernetes cluster. See GitLab's documentation on how to do this properly.

## GitLab Credentials
GitLab has finally introduced a way to have persistent deploy tokens that can fetch things from the Docker registry. k8s-deploy-helper 3.0 now uses this more secure token. You can create a deploy token at Settings->Repository->Deploy Tokens and make one named ```gitlab-deploy-token``` with read_registry access. It HAS to be called ```gitlab-deploy-token``` due to GitLab limitations. Once this token is created, k8s-deploy-helper will pick it up automatically.

# Building Docker Images

Our goal was to make sure Docker containers could be built as quickly as possible without the developers having to micromanage each docker build command on a per-project basis.

Here is a quick example from the .gitlab-ci.yml:

```
build_container:
  stage: build
  script:
    - command build
  only:
    - branches
```

Notice the script only has one command: ```command build``` - k8s-deploy-helper takes it from there, building the container, tagging it with a unique id (commit hash), and pushing it into the GitLab docker registry.

## Caching Docker FS Layers

By default, k8s-deploy-helper and the accompanying examples use a convention where the build that is deployed to production is given the ```latest``` tag after successful deployment. When building Docker containers, we use --cache-from your docker image's ```latest``` tag. This will allow for more optimized caching when using multiple GitLab runners or runners without persistent storage.

If you're managing your own runners, and you only have one, then you may want to think about setting a variable named  ```KDH_SKIP_LATEST``` to ```true``` in your build stages or in the GitLab variables UI. When k8s-deploy-helper finds this variable set to true, we don't use --cache-from, and will just build the Docker container normally, which will try and make use of the cache that is already present in the runner.

## Build Arguments

Sometimes you need to pass in arguments to containers at build time to do things like putting a token in place to pull from a private npm registry. To pass in build arguments, simply go to your GitLab project and go to Settings->CI/CD->Variables and create a variable in this form:

```BUILDARG_npmtoken=1111```

When we build the Docker container, we look for all environment variables that start with ```BUILDARG_```, strip that prefix out, and pass it into docker via --build-arg. In the example above, this will create a build argument ```npmtoken``` with a value of ```1111```

In the example above, you would have to put the following line into your Dockerfile in order to use it at build time:

```ARG npmtoken```

## Automatic ARG Insertion
Starting in 3.0, you can set a variable named ```KDH_INSERT_ARGS``` to ```true```, and k8s-deploy-helper will automatically take all your build arguments and insert corresponding ARG commands into your Dockerfile at build time, immediately after FROM lines. This makes GitLab the source of truth and you as a developer will no longer have to worry about setting things in multiple places.

Inserting things into your Dockerfile at runtime is definitely considered a bit magic though, so we make this a feature you have to opt into.

## Secrets as Buildargs
Sometimes you may have an application that has logic that looks to make sure all environment variables are present before you can run any command (like asset generation). k8s-deploy-helper has the ability to automatically create secrets and use them in your Kubernetes manifests, which we will go into later on. If you set the variable ```BUILDARGS_FROM``` to ```production```, it will take all the secrets that would be created to run in the ```production``` stage and automatically use them as build arguments when creating your Docker container. This will also turn on the ```KDH_INSERT_ARGS``` feature and will insert ```ARG``` statements into your Dockerfile automatically.

## Build Multiple Dockerfiles

If your project needs to build multiple Dockerfiles, the helper will automatically handle all the naming convention management to avoid collisions.  All you need to do is pass in the file name of the Dockerfile that is in the root of your repository. For example, if you have two Dockerfiles, Dockerfile-app, and Dockerfile-worker, this is what your .gitlab-ci.yml would look like:

```
build_app:
  stage: build
  script:
    - command build Dockerfile-app
  only:
    - branches

build_worker:
  stage: build
  script:
    - command build Dockerfile-worker
  only:
    - branches
```

## Buildpack Builds
Starting in 3.1.0, KDH can build applications using [Heroku Buildpacks](https://devcenter.heroku.com/articles/buildpacks) via [herokuish](https://github.com/gliderlabs/herokuish). To do this, we run the latest herokuish docker container to make sure you have access to the latest buildpacks. Because we mount your code into the herokuish docker container, we need to use dind, so you'll need to make sure your GitLab runner has privileged access. All you need to do is not have a Dockerfile in your root, and we'll use the buildpack method. In the gitlab-ci, you'll need to expose docker:stable-dind as a service like so:

```
build:
  stage: build
  services:
    - docker:stable-dind
  script:
    - command build
  only:
    - branches
```

You can use BUILDARG_ syntax from above to pass in build arguments, such as npm tokens, etc...

# Kubernetes Deployment

The key to success here is being able to use variables in your manifests. By using the right variables in the right places, you can have one single deployment manifest to maintain that can create deployments for review apps, staging, canaries, and production. See our [example repository](https://github.com/lifechurch/example-go) for more information on how to properly set up your manifests.

## Directory Structure
To deploy applications into Kubernetes, you need to place your templated manifest files into a ```kubernetes``` directory at the root of your repository.

```
kubernetes
|-->deployment.yaml
|-->service.yaml
```

## Per-Stage Directory Structure

Sometimes you have manifests that you only want to run in particular stages. For instance, you may want horizontal pod autoscalers only for production, but not for staging or review apps. All you have to do is create a directory name that corresponds to your build stage.

```
kubernetes
|production
||-->hpa.yaml
|-->deployment.yaml
|-->service.yaml
```

## Debugging
Starting in 3.0, k8s-deploy-helper renders all templates before trying to apply them. All rendered manifests are displayed in the runner for easy debugging. The rendered templates are put in ```/tmp/kubernetes``` if you want to grab them using GitLab artifacts for some reason. In addition, we now use kubeval.

## kubeval
KDH will try to figure out the version of Kubernetes you are deploying to, and then target kubeval specificaly for that version.

Optionally, you can set the KDH_SKIP_KUBEVAL variable to true in order to skip the use of kubeval.

## Canary Deploys
As of 3.0, k8s-deploy-helper will automatically support canary deployments via rewriting your deployment manifests. To use this functionality, you just need to be in a GitLab CI stage named canary, and k8s-deploy-helper will search for manifests where the ```track``` label is set to ```stable```.

The canary stage operates as a production deployment.

Check out our (example repo)[https://www.github.com/lifechurch/example-go] to see how to set up your manifests to support this automation.

## Environment Variable Substitution
As of 7.0, k8s-deploy-helper added some logic that will only substitute environment variables that exist into your manifest files.  

# Secret Management

For people just getting started with deploying apps to Kubernetes, one of the first questions is 'how do I keep secrets out of my repositories?' k8s-deploy-helper has built-in secret management that allows you to securely use GitLab as the source of truth for all secrets.

How k8s-deploy-helper handles secrets is probably the hardest part to wrap your minds around initially, so read these documents carefully.


## Secret Creation
To create a secret, go to your GitLab project and go to Settings->CI/CD->Variables and create a variable with this name pattern:

```SECRET_mykeyname```

During deployment, our scripts will look for all environment variables that start with the prefix ```SECRET_```, strip out the prefix and sticks the key and value into a kubernetes secret named ```$KUBE_NAMESPACE-secrets-$STAGE```, which translates to something like ```yournamespacename-secrets-production``` or ```yournamespacename-secrets-staging```

In the example above, there would be an entry in the secret file named ```mykeyname``` with the corresponding value you put in GitLab. You can then access these secrets in your manifest files. The below will create an environment variable in your pod called mykeyname.

```
        env:
          - name: mykeyname
            valueFrom:
              secretKeyRef:
                name: $KUBE_NAMESPACE-secrets-$STAGE
                key: mykeyname
```
**The important thing to note is that k8s-deploy-helper does the stripping of the SECRET_ prefix during secret creation TO RUN IN KUBERNETES. When dealing with stages outside of k8s-deploy-helper, like for instance, a testing stage or a stage that does database migrations, your variables are sent as is to your GitLab Runners, prefixes and all.**

## Per-Stage Secret Creation

Sometimes you have secrets that have different values depending on if you're running in production or staging. Our helper allows you to do this by prefixing your secret with the uppercased version of your GitLab CI stage name.

For example, let's say you have a secret called ```api_env```, that needs to have different values depending on if you're deploying to one of three stages: review, staging, or production.

Instead of creating a variable in GitLab called ```SECRET_api_env```, you would create three:

```
REVIEW_api_env
STAGING_api_env
PRODUCTION_api_env
```

Combined with a templated section like below, this would pull in the secret from wherever.

```
        env:
          - name: api_env
            valueFrom:
              secretKeyRef:
                name: $KUBE_NAMESPACE-secrets-$STAGE
                key: api_env
      imagePullSecrets:
```
## Automated Secret Management in manifests

New in version 3.0 is the {{SECRETS}} command. In the examples above, we gave the code that you would insert into manifests to make the secrets that k8s-deploy-helper creates in Kubernetes usable from within your deployments. This meant for adding a new secret, you would have to set the value of the secret in GitLab, and then add some code to the manifest to make it accessible.

Wanting to make developers lives easier and make GitLab the source of truth, we introduced the {{SECRETS}} command that you can insert into your templates at the appropriate place, and when we render your manifest templates, we will loop through all the secrets that k8s-deploy-helper created on your behalf, and insert the appropriate code into the manifest for you!

To use it, just stick {{SECRETS}} in the right place underneath your env: section. Make sure it's placed correctly.

```
env:
  {{SECRETS}}
```

# Deploy Events

Currently NewRelic, Slack, and Datadog deploy events are supported.

In Gitlab for NewRelic, you'll need to add a secret variable with the NewRelic API key and App Ids
for each stage you want a deployment event for. Like:

```
NEWRELIC_API_KEY=xxx
NEWRELIC_STAGING_APP_ID=xxx
NEWRELIC_PRODUCTION_APP_ID=xxx
```

For Slack, you can simply set a Gitlab secret variable with the [Slack webhook URL](https://api.slack.com/incoming-webhooks) if you want notifications for every stage.

```
SLACK_WEBHOOK=xxx
```

If you want notifications for specific stages use the following format.

```
SLACK_{{STAGE}}_WEBHOOK
```

E.g.

```
SLACK_STAGING_WEBHOOK=xxx
SLACK_PRODUCTION_WEBHOOK=xxx
```

For Datadog, you *must* set your Datadog API key with:

```
DATADOG_API_KEY=xxx
```

Optionally, you may set an [app key, message text, and tags to send to Datadog.] (https://docs.datadoghq.com/api/?lang=bash#post-an-event)

The text attribute supports markdown. [This help article](https://help.datadoghq.com/hc/en-us/articles/204778779)
best explains how to add markdown text to the deploy event.

The DATADOG_TAGS variable can be used to send one or more tags with the event. Because this is an
array in the POST, you *must* include quotes around each value. Multiple tags should then be
separated by commas.

```
DATADOG_APP_KEY=xxx
DATADOG_TAGS="deploys:api","foo:bar"
DATADOG_TEXT=\n%%%\n### Success\n%%%
```

# Manifest-less Deploys
Starting in 3.1.0, we added an option for manifest-less deploys to help us migrate away from Deis Workflow. In order for this to work, we had to make some very opinionated decisions regarding our manifests. These may not work for your organization. If this is the case, we encourage you to fork our project and make your own default manifests. They can be found in the manifests directory.

## Manifest-less Requirements

* nginx Ingress Controller

* cert-manager or kube-lego that can issue "Let's Encrypt" certificates via the ```kubernetes.io/tls-acme: 'true'``` annotation.

## Conventions

* We will obey Procfiles and every line will get its own deployment. Web will get an ingress, service, pod disruption budget, and autoscaling. Every other line will be treated as worker, and will just get autoscaling.

## Variables & Defaults

### Web

* LIMIT_CPU: 1 - CPU Resource Limit

* LIMIT_MEMORY: 512Mi - Memory Resource Limit

* SCALE_REPLICAS (Production Only): Not Set - If SCALE_REPLICAS is set, SCALE_MIN and SCALE_MAX will be set to the value of SCALE_REPLICAS.

* SCALE_MIN (Production Only): 2 - Minimum amount of running pods set in the HPA

* SCALE_MAX (Production Only): 4 - Maximum amount of running pods set in the HPA

* SCALE_CPU (Production Only): 60 - CPU usage at which autoscaling occurs

* PDB_MIN (Production Only): 50% - Minimum available percentage

* PORT: 5000 - The port your app listens on

* PROBE_URL: / - The URL that will get hit for readiness probe

* LIVENESS_PROBE: /bin/true - The command used for the liveness probe

### Other (workers)

To set variables for your other runtimes specified in the Procfile, you can create variables with this pattern. For example, let's say you have a worker that's named ```worker``` in your Procfile and you want to assign 2 CPU to each pod, you would set a variable named ```worker_LIMIT_CPU``` to ```2```.

Variables you can set to control your worker stages are listed below, along with their default values. We'll refer to the name of your stage as ${1}.

```
  ${1}_LIMIT_CPU: 1
  ${1}_LIMIT_MEMORY: 512Mi
  ${1}_LIVENESS_PROBE: /bin/true
  ${1}_REPLICAS: Not Set
  ${1}_SCALE_MIN: 1
  ${1}_SCALE_MAX: 1
  ${1}_SCALE_CPU : 60%
```


# Contributing

Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests.

# Versioning
To make sure the community can use this project with their sanity intact, we will be committing to incrementing major versions when we introduce breaking changes.  We anticipate this happening frequently, as this tool is still under heavy development.

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/lifechurch/k8s-deploy-helper/tags).

# License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details.


CHANGELOG.md
# Version 7.1.0

Docker Image: quay.io/lifechurch/k8s-deploy-helper:7.1.0

## Feature

* Adds a deploy-flagger command for Flagger canary deployments

## Bugfixes

* Update destroy command to include configmaps and certificates

# Version 7.0.2

Docker Image: quay.io/lifechurch/k8s-deploy-helper:7.0.2

## Bugfixes

* Herokuish build process couldn't build successfully unless at least one environment variable was set.

# Version 7.0.1

Docker Image: quay.io/lifechurch/k8s-deploy-helper:7.0.1

## Updates

* Updated helm to 2.14.1

# Version 7.0.0

Docker Image: quay.io/lifechurch/k8s-deploy-helper:7.0.0

## Backwards Incompatible Changes

* Depracated use of ${DOLLAR} in manifests in favor of telling envsubst to only substitute environment variables that exist in the CI run. If you previously used ${DOLLAR}, you can just remove the {DOLLAR} portion and it should work fine.

The following command (on macOS) should do the replacement for you: ```find kubernetes -type f -name '*.yaml' | xargs -I {} sed -i '' -e "s/\${DOLLAR}/$/g" "{}"```

## Feature

* Datadog notification text is more easily searchable. You can now overlay KDH deployment events by doing a string like "KDH Deployment: $NAMESPACE production", substituting $NAMESPACE for your namespace.

# Version 6.0.3

Docker Image: quay.io/lifechurch/k8s-deploy-helper:6.0.3

## Updates

* Updated kubeval to 0.12.0

## Bugfixes

* kubeval will skip missing schemas now

# Version 6.0.0

Docker Image: quay.io/lifechurch/k8s-deploy-helper:6.0.0

## Updates

* Updated docker dind image to docker:18.06.3-ce-dind

* Updated Helm to 2.13.1

* Updated kubectl to 1.12.6

* Updated kubeval to 0.8.1

# Version 5.0.1

Docker Image: quay.io/lifechurch/k8s-deploy-helper:5.0.1

## Bug Fixes

* Fixed race condition in {{SECRETS}} substitution.

* Added additional debugging to trap bash errors.

# Version 5.0.0

Docker Image: quay.io/lifechurch/k8s-deploy-helper:5.0.0

## Feature

* DEPLOY: KDH will determine the Kubernetes version you are deploying against and run kubeval manifests specifically against that version.

* DEPLOY: You can now set KDH_SKIP_KUBEVAL variable to true in order to skip kubeval runs.

# Version 4.0.3

Docker Image: quay.io/lifechurch/k8s-deploy-helper:4.0.3

## Feature

* BUILD: Docker builds now pass the --pull flag in case you build from an image that pushes different content to the same tag.

# Version 4.0.2

Docker Image: quay.io/lifechurch/k8s-deploy-helper:4.0.2

## Bug Fixes

* Version 4.0.1 introduced a bug where deployment rollouts were not checked correctly. This is fixed.

## New Features

* New Relic deployment notifications now use the v2 API

* Deployment notifications have more contextual information about what was deployed.

# Version 4.0.1 - DO NOT USE

Docker Image: quay.io/lifechurch/k8s-deploy-helper:4.0.1

## Bug Fixes

* Canary stage was deploying manifests without ```track: stable``` label. This shouldn't happen. Fixed.

* KDH no longer tries to create the namespace if it exists already. This was causing issues because GitLab is now supposed to be creating the namespace when you add the GitLab integration.

# Version 4.0.0

Docker Image: quay.io/lifechurch/k8s-deploy-helper:4.0.0

## Backwards Incompatible Changes

* Starting in GitLab 11.5, GitLab decided to start managing service accounts for namespaces automatically. This broke k8s-deploy-helper because KUBE_TOKEN was no longer the token you specified in the cluster. Moving forward, k8s-deploy-helper is not going to generate a Kubernetes configuration itself, it will rely on GitLab to create it for us.

## New Features

* Slack Deploy Events now support separate notifications per stage, allowing you to only send deploy events on specific stages if desired, or to different Slack channels for each stage.

# Version 3.1.1

Docker Image: quay.io/lifechurch/k8s-deploy-helper:3.1.1

## Bug Fixes

* BUILDARGS_FROM now works with buildpack builds.

* BUILDARGS_FROM is less chatty.


# Version 3.1.0

Docker Image: quay.io/lifechurch/k8s-deploy-helper:3.1.0

**NOTE: Starting with this version, your GitLab runners need to run in privileged mode to allow for Heroku buildpack builds, and container scanning that will come in the next release.**

## New Features

* BUILD - If no Dockerfile is present in the root directory, we will attempt to build an image using Heroku Buildpacks via herokuish. 

* DEPLOY - If no kubernetes folder is present in the root directory, we will attempt to use our own default manifests. This has Procfile support, allowing you to run workers as well. See documentation for more details.

## Bug Fixes

* 3.0.1 introduced a template evaluation error, fixed.

* Slack and Datadog will now announce canary deploys as canary, rather than production.

# Version 3.0.1 - DO NOT USE

## Changes

* Kubeval will not evaluate ```cloud.google.com``` schemas correctly, so we skip these now. 

# Version 3.0.0

## Backwards Incompatible Changes
* kubectl - Due to a bug in kubectl, you'll need to delete the gitlab-registry secret in your namespace before you deploy. Make sure to do it right before you do a deploy to make it non-impactful. To do this, run ```kubectl delete secret gitlab-registry -n=yournamespace```.

* Deploy Token Usage - Previous versions used an actual GitLab username because GitLab didn't have persistent deploy tokens until recently. Now that this feature is in GitLab, we're going to stop using the shared credentials as this is much more secure. Create a deploy token at Settings->Repository->Deploy Tokens and make one named gitlab-deploy-token with read_registry access. As long as it's named gitlab-deploy-token, that's all you should have to do.

* Canary Usage - Delete your canary manifest templates. We will create them automatically now. Make sure track: stable is present as labels in deployments you want to go out in the canary stage.

## New Features
* DEPLOY - Canary manifests are now dynamically created in canary stages.

src/common.bash
set -eo pipefail

[[ "$TRACE" ]] && set -x

export CI_CONTAINER_NAME="ci_job_build_$CI_JOB_ID"
export CI_REGISTRY_TAG="$CI_COMMIT_SHA"

if [[ "$CI_JOB_STAGE" == "review" ]]; then
  export STAGE="$CI_ENVIRONMENT_SLUG"
else
  export STAGE="$CI_JOB_STAGE"
fi

if [[ "$CI_JOB_STAGE" == "canary" ]]; then
  export STAGE="production"
  export CI_JOB_STAGE="production"
  export ISCANARY="true"
  export REAL_JOB_STAGE="canary"
  echo "We are in a canary deploy, behaving like production but looking for track:"
else
  export ISCANARY="false"
  export REAL_JOB_STAGE=$CI_JOB_STAGE
fi

ensure_deploy_variables() {
    if [[ -z "$KUBE_URL" ]]; then
      echo "ERROR: Missing KUBE_URL. Make sure to configure the Kubernetes Cluster in Operations->Kubernetes"
      exit 1
    fi
    if [[ -z "$KUBE_TOKEN" ]]; then
      echo "ERROR: Missing KUBE_TOKEN. Make sure to configure the Kubernetes Cluster in Operations->Kubernetes"
      exit 1
    fi
    if [[ -z "$KUBE_NAMESPACE" ]]; then
      echo "ERROR: Missing KUBE_NAMESPACE. Make sure to configure the Kubernetes Cluster in Operations->Kubernetes"
      exit 1
    fi
    if [[ -z "$CI_ENVIRONMENT_SLUG" ]]; then
      echo "ERROR: Missing CI_ENVIRONMENT_SLUG. Make sure to configure the Kubernetes Cluster in Operations->Kubernetes"
      exit 1
    fi
    if [[ -z "$CI_ENVIRONMENT_URL" ]]; then
      echo "ERROR: Missing CI_ENVIRONMENT_URL. Make sure to configure the Kubernetes Cluster in Operations->Kubernetes"
      exit 1
    fi
    if [[ -z "$CI_DEPLOY_USER" ]]; then
      echo "ERROR: Missing CI_DEPLOY_USER. Create a deploy token at Settings->Repository->Deploy Tokens and make one named gitlab-deploy-token with read_registry access."
      exit 1
    fi
    if [[ -z "$CI_DEPLOY_PASSWORD" ]]; then
      echo "ERROR: Missing CI_DEPLOY_PASSWORD. Create a deploy token at Settings->Repository->Deploy Tokens and make one named gitlab-deploy-token with read_registry access."
      exit 1
    fi
}

ping_kube() {
  if kubectl version > /dev/null; then
    echo "Kubernetes is online!"
    echo ""
  else
    echo "Cannot connect to Kubernetes."
    return 1
  fi
}

buildargs_from() {
  if [[ -n "$BUILDARGS_FROM" ]]; then
    echo "BUILDARGS_FROM is set to $BUILDARGS_FROM, starting clone secret operation."
    echo "Turning on the KDH_INSERT_ARGS flag"
    export KDH_INSERT_ARGS=true
    if env | grep -i -e '^SECRET_' > /dev/null; then
      IFS=$'\n'
      SECRETS=$(env | grep -i -e '^SECRET_')
      for i in $SECRETS; do
        fullkey=$(echo $i | cut -d'=' -f1)
        stripped=$(echo $i | cut -d'_' -f2-)
        key=$(echo $stripped | cut -d'=' -f1)
        value=$(echo -n "${!fullkey}")
        echo "Exporting $key as BUILDARG_$key"
        export BUILDARG_$stripped
      done
    fi
    if env | grep -i -e "^${BUILDARGS_FROM}_" > /dev/null; then
      IFS=$'\n'
      STAGE_SECRETS=$(env | grep -i -e "^${BUILDARGS_FROM}_")
      for i in $STAGE_SECRETS; do
        fullkey=$(echo $i | cut -d'=' -f1)
        stripped=$(echo $i | cut -d'_' -f2-)
        key=$(echo $stripped | cut -d'=' -f1)
        value=$(echo -n "${!fullkey}")
        echo "Exporting $key as BUILDARG_$key"
        export BUILDARG_$stripped
      done
    fi
  fi
}

insert_args() {
  if [[ -n "$KDH_INSERT_ARGS" ]]; then
    echo "KDH_INSERT_ARGS is turned on, so we're going to re-write your Dockerfile and insert ARG commands for every BUILDARG"
    IFS=$'\n'
    ALL_VARIABLES=$(env | grep -i -e '^BUILDARG_')
    for i in $ALL_VARIABLES; do
      stripped=$(echo $i | cut -d'_' -f2-)
      key=$(echo $stripped | cut -d'=' -f1)
      echo "Inserting ARG $key into Dockerfile below the FROM"
      sed -i -e "/^FROM/a ARG $key" $DOCKERFILE
    done
    echo "Dockerfile manipulation complete. Now it looks like:"
    echo
    cat $DOCKERFILE
    echo 
  fi
}

set_defaults() {

  if [[ -v SCALE_REPLICAS ]]; then
    export SCALE_MIN=$SCALE_REPLICAS
    export SCALE_MAX=$SCALE_REPLICAS
  fi

  if [[ ! -v SCALE_MIN ]]; then
    export SCALE_MIN=2
  fi

  if [[ ! -v SCALE_MAX ]]; then
    export SCALE_MAX=4
  fi

  if [[ ! -v SCALE_CPU ]]; then
    export SCALE_CPU=60
  fi

  if [[ ! -v PDB_MIN ]]; then
    export PDB_MIN="50%"
  fi

  if [[ ! -v PORT ]]; then
    export PORT=5000
  fi

  if [[ ! -v PROBE_URL ]]; then
    export PROBE_URL="/"
  fi

  if [[ ! -v LIMIT_CPU ]]; then
    export LIMIT_CPU="1"
  fi

  if [[ ! -v LIMIT_MEMORY ]]; then
    export LIMIT_MEMORY="512Mi"
  fi

  if [[ ! -v LIVENESS_PROBE ]]; then
    export LIVENESS_PROBE="/bin/true"
  fi
}

set_prefix_defaults() {
  memory=${1}_LIMIT_MEMORY
  cpu=${1}_LIMIT_CPU
  liveness=${1}_LIVENESS_PROBE
  replicas=${1}_REPLICAS
  min_replicas=${1}_SCALE_MIN
  max_replicas=${1}_SCALE_MAX
  scale_cpu=${1}_SCALE_CPU

  if [[ -v ${replicas} ]]; then
    export ${min_replicas}=${!replicas}
    export ${max_replicas}=${!replicas}
  fi

  if [[ ! -v ${min_replicas} ]]; then
    export ${min_replicas}="1"
  fi

  if [[ ! -v ${max_replicas} ]]; then
    export ${max_replicas}="1"
  fi

  if [[ ! -v ${scale_cpu} ]]; then
    export ${scale_cpu}="60"
  fi

  if [[ ! -v ${memory} ]]; then
    export ${memory}="512Mi"
  fi

  if [[ ! -v ${cpu} ]]; then
    export ${cpu}="1"
  fi

  if [[ ! -v ${liveness} ]]; then
    export ${liveness}="/bin/true"
  fi
}

set_buildargs() {
  IFS=$'\n'
  if env | grep -i -e '^BUILDARG_' > /dev/null; then
    ALL_VARIABLES=$(env | grep -i -e '^BUILDARG_')
    for i in $ALL_VARIABLES; do
      fullkey=$(echo $i | cut -d'=' -f1)
      stripped=$(echo $i | cut -d'_' -f2-)
      key=$(echo $stripped | cut -d'=' -f1)
      value=$(echo -n "${!fullkey}")
      buildargs="${buildargs}--build-arg $key='$value' "
    done
    export buildargs=$buildargs
  fi
}

build_env() {
  IFS=$'\n'
  echo "Removing .env file"
  rm $CI_PROJECT_DIR/.env &> /dev/null || true
  if env | grep -i -e '^BUILDARG_' > /dev/null; then
    ALL_VARIABLES=$(env | grep -i -e '^BUILDARG_')
    for i in $ALL_VARIABLES; do
      fullkey=$(echo $i | cut -d'=' -f1)
      stripped=$(echo $i | cut -d'_' -f2-)
      key=$(echo $stripped | cut -d'=' -f1)
      value=$(echo -n "${!fullkey}")
      echo "$key=$value" >> $CI_PROJECT_DIR/.env
    done
  fi
}

setup_docker() {
  if ! docker info &>/dev/null; then
    if [ -z "$DOCKER_HOST" -a "$KUBERNETES_PORT" ]; then
      export DOCKER_HOST='tcp://localhost:2375'
    fi
  fi
}

get_secrets_for_creation() {
  if env | grep -i -e '^SECRET_' > /dev/null; then
    IFS=$'\n'
    SECRETS=$(env | grep -i -e '^SECRET_')
    for i in $SECRETS; do
      fullkey=$(echo $i | cut -d'=' -f1)
      stripped=$(echo $i | cut -d'_' -f2-)
      key=$(echo $stripped | cut -d'=' -f1)
      value=$(echo -n "${!fullkey}" | base64 -w 0)
      echo "  $key: $value"
    done
  fi

  if env | grep -i -e "^$CI_JOB_STAGE"_ > /dev/null; then
    IFS=$'\n'
    STAGE_SECRETS=$(env | grep -i -e "^$CI_JOB_STAGE")
    for i in $STAGE_SECRETS; do
      fullkey=$(echo $i | cut -d'=' -f1)
      stripped=$(echo $i | cut -d'_' -f2-)
      key=$(echo $stripped | cut -d'=' -f1)
      value=$(echo -n "${!fullkey}" | base64 -w 0)
      echo "  $key: $value"
    done
  fi
}

get_secrets_for_usage() {
  if env | grep -i -e '^SECRET_' > /dev/null; then
    IFS=$'\n'
    SECRETS=$(env | grep -i -e '^SECRET_')
    for i in $SECRETS; do
      fullkey=$(echo $i | cut -d'=' -f1)
      stripped=$(echo $i | cut -d'_' -f2-)
      key=$(echo $stripped | cut -d'=' -f1)
      value=$(echo -n "${!fullkey}" | base64 -w 0)
      echo "- name: $key" >> /tmp/secrets.yaml
      echo "  valueFrom:" >> /tmp/secrets.yaml
      echo "    secretKeyRef:" >> /tmp/secrets.yaml
      echo "      name: $KUBE_NAMESPACE-secrets-$STAGE" >> /tmp/secrets.yaml
      echo "      key: $key" >> /tmp/secrets.yaml
    done
  fi

  if env | grep -i -e "^$CI_JOB_STAGE"_ > /dev/null; then
    IFS=$'\n'
    STAGE_SECRETS=$(env | grep -i -e "^$CI_JOB_STAGE")
    for i in $STAGE_SECRETS; do
      fullkey=$(echo $i | cut -d'=' -f1)
      stripped=$(echo $i | cut -d'_' -f2-)
      key=$(echo $stripped | cut -d'=' -f1)
      value=$(echo -n "${!fullkey}" | base64 -w 0)
      echo "- name: $key" >> /tmp/secrets.yaml
      echo "  valueFrom:" >> /tmp/secrets.yaml
      echo "    secretKeyRef:" >> /tmp/secrets.yaml
      echo "      name: $KUBE_NAMESPACE-secrets-$STAGE" >> /tmp/secrets.yaml
      echo "      key: $key" >> /tmp/secrets.yaml
    done
  fi
}

get_deploy_events() {
  if env | grep -i -e '^NEWRELIC_' > /dev/null; then
    NEWRELIC=$(env | grep -i -e '^NEWRELIC_')
    for i in $NEWRELIC; do
        if echo "$i" | grep -i -e "API_KEY" > /dev/null; then
            export NEWRELIC_API_KEY=$(echo $i | cut -d'=' -f2)
        fi
        if echo "$i" | grep -i -e "$CI_JOB_STAGE" | grep -i -e "APP_ID" > /dev/null; then
            export NEWRELIC_APP_ID=$(echo $i | cut -d'=' -f2)
        fi
    done
  fi

  if env | grep -i -e '^SLACK' > /dev/null; then
    SLACK=$(env | grep -i -e '^SLACK')
    for i in $SLACK; do
        if echo "$i" | grep -i -e "$CI_JOB_STAGE" | grep -i -e "WEBHOOK" > /dev/null; then
            export SLACK_WEBHOOK_URL=$(echo $i | cut -d'=' -f2)
        elif echo "$i" | grep -i -e "^SLACK_WEBHOOK" > /dev/null; then
            export SLACK_WEBHOOK_URL=$(echo $i | cut -d'=' -f2)
        fi
    done
  fi
}